chain-signatures/contract/src/lib.rs

```diff
@@ -533,15 +501,74 @@ impl VersionedMpcContract {
                     Ok(true)
                 } else {
                     Err(MpcContractError::VoteError(
-                        VoteError::UnexpectedProtocolState("resharing".to_string()),
+                        VoteError::UnexpectedProtocolState("Running".to_string()),
                     ))
                 }
             }
             _ => Err(MpcContractError::VoteError(
-                VoteError::UnexpectedProtocolState("resharing".to_string()),
+                VoteError::UnexpectedProtocolState("Running".to_string()),
             )),
         }
     }
+
+    /// Propose an update to the contract. [`Update`] are all the possible updates that can be proposed.
+    ///
+    /// returns Some(id) if the proposal was successful, None otherwise
+    #[handle_result]
+    pub fn propose_update(
+        &mut self,
+        code: Option<Vec<u8>>,
+        config: Option<Config>,
+    ) -> Result<UpdateId, VoteError> {
+        // Only voters can propose updates:
+        self.voter()?;
+
+        let Some(id) = self.proposed_updates().propose(code, config) else {
+            return Err(VoteError::Unexpected(
+                "cannot propose update due to incorrect parameters".into(),
+            ));
+        };
+
+        env::log_str(&format!("id={id:?}"));
+        Ok(id)
+    }
+
+    /// Vote for a proposed update given the [`UpdateId`] of the update.
+    ///
+    /// Returns Ok(true) if the amount of participants surpassed the threshold and the update was executed.
+    #[handle_result]
+    pub fn vote_update(&mut self, id: UpdateId) -> Result<bool, MpcContractError> {
+        let threshold = match self {
+            Self::V0(contract) => match &contract.protocol_state {
+                ProtocolContractState::Running(state) => state.threshold,
+                ProtocolContractState::Resharing(state) => state.threshold,
+                _ => {
+                    return Err(MpcContractError::VoteError(
+                        VoteError::UnexpectedProtocolState("Initialized or NotInitialized".into()),
+                    ))
+                }
+            },
+        };
+
+        let voter = self.voter()?;
+        let Some(votes) = self.proposed_updates().vote(&id, voter) else {
+            return Err(MpcContractError::VoteError(VoteError::UpdateNotFound));
+        };
+
+        // Not enough votes, wait for more.
+        if votes.len() < threshold {
+            return Ok(false);
+        }
+
+        let Some(_promise) =
+            self.proposed_updates()
+                .do_update(&id, "update_config", UPDATE_CONFIG_GAS)
```

Comment:
This looks a bit odd. You are passing "update_config", but this function will both update the config and do the update.